tests/cpp/test_multidevice_tutorial.cpp

```diff
@@ -375,4 +371,251 @@ TEST_F(MultiDeviceTutorial, SimplePipelining) {
+// axis is parallelized with DIDx. It means that each device will materialize a
+// tensor of shape {1,128} representing a slice of the global tensor.
+
+// Note 1: the extent of the DeviceMesh always needs to match the tensor's axis
+// extent. This is similar to the fact that, in Fuser single device, if an axis
+// of extent 32 is parallelized with TIDx, then the launch param blockDim
+// (i.e. the number of threads per blocks) will be chosen equal to 32.
+
+// Note 2: sharding a tensor (i.e. parallelizing an axis with DIDx) is
+// meaningless without an underlying device mesh, because otherwise we cannot
+// guess on what devices the tensor needs to be sharded. This is in apparence
+// different to BIDx and TIDx parallel, where we don't access the physical
+// threads and blocks to map to.
+TEST_F(MultiDeviceTutorial, TensorShardingAndResharding) {
+  // MODEL DEFINITION
+  // Let us define a model expressing a simple memcpy
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* tv0 = makeContigTensor(2);
+  fusion->addInput(tv0);
+  TensorView* tv1 = set(tv0); // "set" means "copy"
+  fusion->addOutput(tv1);
+
+  // MULTIDEVICE SCHEDULING
+  // Let us define, as in previous tests, a 1D Device Mesh comprised of all
+  // available device IDs
+  std::vector<int64_t> all_devices(communicator_->size());
+  std::iota(
+      all_devices.begin(),
+      all_devices.end(),
+      0); // all_devices = [0,1,..., communicator_->size()-1]
+  DeviceMesh mesh_full(all_devices);
+  // Let us set tv0 and tv1's mesh:
+  tv0->setDeviceMesh(mesh_full);
+  tv1->setDeviceMesh(mesh_full);
+  // Without further parallelization, this means that the tensors are fully
+  // replicated on all the devices. Let us now introduce some parallelization to
+  // describe how the tensors are sharded onto this device mesh.
+
+  // ################################
+  // Fully sharded with no resharding
+  // ################################
+  // Let us shard the tensors' outermost axis onto the device mesh:
+  tv0->axis(0)->parallelize(ParallelType::DIDx);
+  tv1->axis(0)->parallelize(ParallelType::DIDx);
+  // It means that each device will own a slice of the the full tensors. In
+  // particular, it implies that tv0's and tv1's outermost axis extent equals
+  // the number of devices. The shapes of tv0 and tv1 are thus
+  // [communicator_->size(), ?]
+  const bool verbose_print = verbose_ && communicator_->deviceId() == 0;
+  if (verbose_print) {
+    std::cout << "tv0: " << tv0->toString() << std::endl;
+    std::cout << "tv1: " << tv1->toString() << std::endl;
+  }
+  // However, the outermost axis with extent `communicator_->size()`is not
+  // materialized on one device, but accross devices, therefore, each device
+  // allocate a tensor of shape {1, ?}, representing a slice of the global
+  // tensor.
+
+  // RUNTIME
+  // Set up the input
+  constexpr int64_t tensor_size = 128;
+  const c10::TensorOptions tensor_options =
```

Comment:
```suggestion
  const auto tensor_options =
```
because `at::TensorOptions()` spells out the type already.